torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+              graph_->insertConstant(static_cast<int64_t>(*maybe_len)));
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = boolean_value_refinements_.count(if_n.cond()) != 0;
+        ListRefinement empty;
+        auto true_block_refinements = RefineListLens(
+            if_n.thenBlock(),
+            has_cond_ref ? boolean_value_refinements_[if_n.cond()].true_refine()
+                         : empty);
+        auto false_block_refinements = RefineListLens(
+            if_n.elseBlock(),
+            has_cond_ref
+                ? boolean_value_refinements_[if_n.cond()].false_refine()
+                : empty);
+        bool true_block_throws = throwing_blocks_.count(if_n.thenBlock());
+        bool false_block_throws = throwing_blocks_.count(if_n.elseBlock());
+
+        // if one block throws, the refinements for the other block
+        // become present in the current block, and all bool outputs
+        // of the if node take their refinements from non throwing block
+        // output
+
+        if (true_block_throws || false_block_throws) {
+          if (true_block_throws && false_block_throws) {
+            throwing_blocks_.insert(b);
+            continue;
+          }
+          if (true_block_throws) {
+            block_refinements.insert(
+                false_block_refinements.begin(), false_block_refinements.end());
+          } else {
+            block_refinements.insert(
+                true_block_refinements.begin(), true_block_refinements.end());
+          }
+          Block* non_throwing_block =
+              true_block_throws ? n->blocks().at(1) : n->blocks().at(0);
+          for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+            if (boolean_value_refinements_.count(
+                    non_throwing_block->outputs().at(i))) {
+              boolean_value_refinements_[n->outputs().at(i)] =
+                  boolean_value_refinements_[non_throwing_block->outputs().at(
+                      i)];
+            }
+          }
+          continue;
+        }
+
+        for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+          if (!(if_n.outputs().at(i)->type() == BoolType::get())) {
+            continue;
+          }
+          Value* true_v = if_n.thenOutputs().at(i);
+          Value* false_v = if_n.elseOutputs().at(i);
+
+          if (!boolean_value_refinements_.count(true_v) &&
+              !boolean_value_refinements_.count(false_v) &&
+              !constant_as<bool>(true_v) && !constant_as<bool>(false_v)) {
+            continue;
+          }
+
+          // if either block has a constant bool output, e.g. `true` on the
+          // true block, then for the `false` value we can take the false
+          // refinements present on the false block and from the other block
+          // output value bc if the output is false it had to have come from the
+          // false block. if len(a) == 5:
```

Comment:
Nit: start example from a new line.

Nit2: the description goes about a case where we have a true on the true branch, but example illustrates false on false branch. It's probably better to use the same case.

Nit3: might be worth writing examples in torch-scripts, bc in python true_v and false_v are both "x", while in TS they would be explicitly different.